File: reference-implementations/python/oh_my_loop/feedback.py
"""Optional feedback store - persists loop outcomes across sessions.

This is OPTIONAL. Skills are stateless by default; this module provides
a simple JSON store for users who want feedback accumulation.
"""
import json
import os
import tempfile
import threading
from datetime import datetime, timezone
from typing import Optional


class FeedbackStore:
    """Simple JSON file store for loop outcomes. Optional component."""

    def __init__(self, path: str = "feedback.json", *, consent: bool = False, max_entries: int = 500):
        if max_entries < 1:
            raise ValueError("max_entries must be >= 1")
        self.path = path
        self.consent = consent
        self.max_entries = max_entries
        self._lock = threading.RLock()

    def record(self, task: str, pattern: str, success: bool, cost_tokens: int, failure_mode: Optional[str] = None):
        if not self.consent:
            raise PermissionError("feedback persistence requires explicit consent")
        entry = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "task": task,
            "pattern": pattern,
            "success": success,
            "cost_tokens": cost_tokens,
            "failure_mode": failure_mode,
        }
        history = self._read()
        history.append(entry)
        self._write(history[-self.max_entries :])

    def get_recent(self, n: int = 20) -> list:
        if n <= 0:
            return []
        return self._read()[-n:]

    def get_failure_rate(self, pattern: Optional[str] = None) -> float:
        history = self._read()
        if pattern:
            history = [h for h in history if h["pattern"] == pattern]
        if not history:
            return 0.0
        failures = sum(1 for h in history if not h["success"])
        return failures / len(history)

    def _read(self) -> list:
        with self._lock:
            if not os.path.exists(self.path):
                return []
            with open(self.path, encoding="utf-8") as f:
                try:
                    data = json.load(f)
                except json.JSONDecodeError as exc:
                    raise ValueError(f"feedback store is corrupted: {exc}") from exc
            if not isinstance(data, list):
                raise ValueError("feedback store must contain a JSON list")
            return data

    def _write(self, data: list):
        with self._lock:
            directory = os.path.dirname(os.path.abspath(self.path))
            os.makedirs(directory, exist_ok=True)
            fd, tmp_path = tempfile.mkstemp(prefix=".feedback.", dir=directory)
            try:
                with os.fdopen(fd, "w", encoding="utf-8") as f:
                    json.dump(data, f, ensure_ascii=False, indent=2)
                    f.flush()
                    os.fsync(f.fileno())
                os.replace(tmp_path, self.path)
            finally:
                if os.path.exists(tmp_path):
                    os.unlink(tmp_path)

File: reference-implementations/typescript/src/feedback.ts
/** Consent-gated local outcome store. Do not use for sensitive life memory. */

import * as fs from "fs";
import * as pathModule from "path";

export interface FeedbackEntry {
  timestamp: string;
  task: string;
  pattern: string;
  success: boolean;
  costTokens: number;
  failureMode?: string;
}

export class FeedbackStore {
  constructor(
    private readonly path: string = "feedback.json",
    private readonly consent: boolean = false,
    private readonly maxEntries: number = 1000,
  ) {}

  record(entry: Omit<FeedbackEntry, "timestamp">): void {
    if (!this.consent) throw new Error("feedback persistence requires explicit consent");
    const history = this.read();
    history.push({ ...entry, timestamp: new Date().toISOString() });
    this.write(history.slice(-this.maxEntries));
  }

  getRecent(n: number = 20): FeedbackEntry[] { return this.read().slice(-Math.max(0, n)); }

  getFailureRate(pattern?: string): number {
    const history = pattern ? this.read().filter((h) => h.pattern === pattern) : this.read();
    if (history.length === 0) return 0;
    return history.filter((h) => !h.success).length / history.length;
  }

  private read(): FeedbackEntry[] {
    if (!fs.existsSync(this.path)) return [];
    const value: unknown = JSON.parse(fs.readFileSync(this.path, "utf-8"));
    if (!Array.isArray(value)) throw new Error("feedback store is corrupt: expected an array");
    return value as FeedbackEntry[];
  }

  private write(data: FeedbackEntry[]): void {
    const directory = pathModule.dirname(pathModule.resolve(this.path));
    fs.mkdirSync(directory, { recursive: true });
    const temp = `${this.path}.${process.pid}.tmp`;
    fs.writeFileSync(temp, JSON.stringify(data, null, 2), { encoding: "utf-8", mode: 0o600 });
    fs.renameSync(temp, this.path);
  }
}
